src/hooks/useForm/FormProvider.js

```diff
@@ -0,0 +1,259 @@
+
+    /** Container styles */
+    style: stylePropTypes,
+
+    /** Custom content to display in the footer after submit button */
+    footerContent: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
+
+    /** Information about the network */
+    network: networkPropTypes.isRequired,
+
+    ...withLocalizePropTypes,
+};
+
+const defaultProps = {
+    isSubmitButtonVisible: true,
+    formState: {
+        isLoading: false,
+    },
+    enabledWhenOffline: false,
+    isSubmitActionDangerous: false,
+    scrollContextEnabled: false,
+    footerContent: null,
+    style: [],
+    validate: () => ({}),
+};
+
+function getInitialValueByType(valueType) {
+    switch (valueType) {
+        case 'string':
+            return '';
+        case 'boolean':
+            return false;
+        case 'date':
+            return new Date();
+        default:
+            return '';
+    }
+}
+
+function FormProvider({validate, shouldValidateOnBlur, shouldValidateOnChange, children, formState, network, enabledWhenOffline, onSubmit, ...rest}) {
+    const inputRefs = useRef({});
+    const touchedInputs = useRef({});
+    const [inputValues, setInputValues] = useState({});
+    const [errors, setErrors] = useState([]);
+
+    const onValidate = useCallback(
+        (values) => {
+            const validateErrors = validate(values);
+            setErrors(validateErrors);
+        },
+        [validate],
+    );
+
+    /**
+     * @param {String} inputID - The inputID of the input being touched
+     */
+    const setTouchedInput = useCallback(
+        (inputID) => {
+            touchedInputs.current[inputID] = true;
+        },
+        [touchedInputs],
+    );
+
+    const submit = useCallback(() => {
+        // Return early if the form is already submitting to avoid duplicate submission
+        if (formState.isLoading) {
+            return;
+        }
+
+        // Touches all form inputs so we can validate the entire form
+        _.each(inputRefs.current, (inputRef, inputID) => (touchedInputs.current[inputID] = true));
+
+        // Validate form and return early if any errors are found
+        if (!_.isEmpty(onValidate(inputValues))) {
+            return;
+        }
+
+        // Do not submit form if network is offline and the form is not enabled when offline
+        if (network.isOffline && !enabledWhenOffline) {
+            return;
+        }
+        // Call submit handler
+        onSubmit(inputValues);
+    }, [enabledWhenOffline, formState.isLoading, inputValues, network.isOffline, onSubmit, onValidate]);
+
+    const registerInput = useCallback(
+        (inputID, propsToParse = {}) => {
+            const newRef = propsToParse.ref || createRef();
+            inputRefs[inputID] = newRef;
+
+            // We want to initialize the input value if it's undefined
+            if (_.isUndefined(inputValues[inputID])) {
+                inputValues[inputID] = propsToParse.defaultValue || getInitialValueByType(propsToParse.valueType);
```

Comment:
I think we can optimize this a bit. Right now, if `propsToParse.defaultValue` is set but is falsy, we'd always run `getInitialValueByType` which would likely return the same value, but it's additional logic we're running without needed to.